lib/compiler/src/beam_ssa_opt.erl

```diff
@@ -2245,6 +2295,176 @@ used_blocks([{L,Blk}|Bs], Def, Acc0) ->
 used_blocks([], _Def, Acc) ->
     rel2fam(Acc).
 
+%% Partition sinks for get_tuple_element instructions in the same
+%% clause extracting from the same tuple. Sort each partition in
+%% execution order.
+partition_deflocs(DefLoc, _Defs, Blocks) ->
+    {BlkNums0,_} = mapfoldl(fun(L, N) -> {{L,N},N+1} end, 0, beam_ssa:rpo(Blocks)),
+    BlkNums = maps:from_list(BlkNums0),
+    S = [{Tuple,{map_get(To, BlkNums),{V,{From,To}}}} ||
+            {V,Tuple,{From,To}} <- DefLoc],
+    F = rel2fam(S),
+    partition_deflocs_1(F, Blocks).
+
+partition_deflocs_1([{Tuple,DefLocs0}|T], Blocks) ->
+    DefLocs1 = [DL || {_,DL} <- DefLocs0],
+    DefLocs = partition_dl(DefLocs1, Blocks),
+    [{Tuple,DL} || DL <- DefLocs] ++ partition_deflocs_1(T, Blocks);
+partition_deflocs_1([], _) -> [].
+
+partition_dl([_]=DefLoc, _Blocks) ->
+    [DefLoc];
+partition_dl([{_,{_,First}}|_]=DefLoc0, Blocks) ->
+    RPO = beam_ssa:rpo([First], Blocks),
+    {P,DefLoc} = partition_dl_1(DefLoc0, RPO, []),
+    [P|partition_dl(DefLoc, Blocks)];
+partition_dl([], _Blocks) -> [].
+
+partition_dl_1([{_,{_,L}}=DL|DLs], [L|_]=Ls, Acc) ->
+    partition_dl_1(DLs, Ls, [DL|Acc]);
+partition_dl_1([_|_]=DLs, [_|Ls], Acc) ->
+    partition_dl_1(DLs, Ls, Acc);
+partition_dl_1([], _, Acc) ->
+    {reverse(Acc),[]};
+partition_dl_1([_|_]=DLs, [], Acc) ->
+    {reverse(Acc),DLs}.
+
+filter_deflocs([{Tuple,DefLoc0}|DLs], Blocks) ->
+    %% Input is a list of sinks of get_tuple_element instructions in
+    %% execution order from the same tuple in the same clause.
+    [{_,{_,First}}|_] = DefLoc0,
+    Paths = find_paths_to_check(DefLoc0, First),
+    WillGC0 = ordsets:from_list([FromTo || {{_,_}=FromTo,_} <- Paths]),
+    WillGC1 = [{{From,To},will_gc(From, To, Blocks, true)} ||
+                  {From,To} <- WillGC0],
+    WillGC = maps:from_list(WillGC1),
+
+    %% Separate sinks that will force the reference to the tuple to be
+    %% saved on the stack from sinks that don't force.
+    {DefLocGC0,DefLocNoGC} =
+        partition(fun({{_,_}=FromTo,_}) ->
+                          map_get(FromTo, WillGC)
+                  end, Paths),
+
+    %% Here is an heuristic to avoid harmful sinking in
+    %% lists:mapfold/3 and similar functions.
+    DefLocGC = case DefLocGC0 of
```

Comment:
This looks a bit heavy, could you break it out into a separate function?

Reply:
Will do.